### ISSUE TITLE
When causes context.Item to lose it's value

### ISSUE THREAD

**Issue by ellinge** [danielgerlag/workflow-core#660]
If you use  When(_ => 0)  instead of When(0) your context.Item loses it's value in the sub tasks.
I'm using it inside a ForEach-step.

When using when with expressions ,context.Item is null
```
.When(_ => 0).Do(then => 
   then.StartWith<DummyStep>()
      .Input(step => step.Value, (_, context) => context.Item)
```

The [Obsolete]-tagged When works fine, a.k.a context.Item is not null
```
.When(0).Then<DummyStep>()
   .Input(step => step.Value, (_, context) => context.Item)
```

**Comment by ellinge** [comment]
If I check inside the Run-step. like here https://github.com/danielgerlag/workflow-core/issues/311. the context.Item is null as well so can't see how I can "extract it" again.

**Comment by ellinge** [comment]
```
using System;
using System.Diagnostics;
using System.Linq;
using SJV.Dirigenten.WebApp.Interface;
using WorkflowCore.Interface;
using WorkflowCore.Models;

namespace ForGithubIssue
{
    public class MultipleApproverWorkflow : IWorkflow<MultipleApproverData>
    {
        public string Id { get; } = "MultipleApproverWorkflow";

        public int Version { get; } = 1;

        public void Build(IWorkflowBuilder<MultipleApproverData> builder)
        {
            builder
                .ForEach(d => d.Approvers.Select((_, i) => i).ToArray())
                    .Do(then =>
                    {
                        var userTask = then.StartWith<SomeUserStep>()
                            .Input(data => data.Prefix, (_, _) => "When(true/false)")
                            .Input(data => data.Index, (_, context) => context.Item)
                            .Input(data => data.Approvers, (data, _) => data.Approvers);

                        userTask.When(true).Then<SomeUserStep>()
                            .Input(data => data.Prefix, (_, _) => "When(true)")
                            .Input(data => data.Index, (_, context) => context.Item)
                            .Input(data => data.Approvers, (data, _) => data.Approvers);

                        userTask.When(false).Then(userTask);
                    })
                .ForEach(d => d.Approvers.Select((_, i) => i).ToArray())
                    .Do(then =>
                    {
                        var userTask = then.StartWith<SomeUserStep>()
                            .Input(data => data.Prefix, (_, _) => "When(_ > true/false)")
                            .Input(data => data.Index, (_, context) => context.Item)
                            .Input(data => data.Approvers, (data, _) => data.Approvers);

                        userTask.When(_ => true).Do(then => then.Then<SomeUserStep>()
                            .Input(data => data.Prefix, (_, _) => "When(_ > true)")
                            .Input(data => data.Index, (_, context) => context.Item)
                            .Input(data => data.Approvers, (data, _) => data.Approvers));
                        userTask.When(_ => false).Do(x => x.Then(userTask));
                    });
        }
    }

    public class SomeUserStep : StepBody
    {
        public string Prefix { get; set; }
        public object Index { get; set; }
        public Approver[] Approvers { get; set; }

        public override ExecutionResult Run(IStepExecutionContext context)
        {
            Approver approver = null;
            if (Index is long index && index >= 0)
                approver = Approvers[index];

            if (approver != null)
            {
                approver.Executions++;
                if (approver.Executions > 1)
                {
                    approver.Responded = true;
                }
            }

            Debug.WriteLine($"{Prefix}: Approver = {approver?.Name} ({approver?.Executions} executions) context.Item = {context.Item}");

            return ExecutionResult.Outcome(approver?.Responded ?? false);
        }
    }

    public class MultipleApproverData : IBaseData
    {
        public Approver[] Approvers { get; set; }
    }

    public class Approver
    {
        public string Name { get; set; }
        public bool Responded { get; set; }
        public int Executions { get; set; } = 0;
    }
}

```

We have a looping task inside the foreach which WaitFor an event from the user/api for a approve/reject/reassign. Everything works great when using the Obselete When but we lose context if we try to migrate to .When(_ =>.
Actually we have two collections of approvers, which can approve in parallel or serial with branching but that's not relevant here.

The response from above when submitting { "Approvers":[ { "Name": "User 1" }, { "Name": "User 2" } ] }: 
> When(true/false): Approver = User 2 (1 executions) context.Item = 1
> When(true/false): Approver = User 1 (1 executions) context.Item = 0
> When(true/false): Approver = User 2 (2 executions) context.Item = 1
> When(true/false): Approver = User 1 (2 executions) context.Item = 0
> When(true): Approver = User 2 (3 executions) context.Item = 1
> When(true): Approver = User 1 (3 executions) context.Item = 0
> When(_ > true/false): Approver = User 2 (4 executions) context.Item = 1
> When(_ > true/false): Approver = User 1 (4 executions) context.Item = 0
> When(_ > true): Approver =  ( executions) context.Item = 
> When(_ > true): Approver =  ( executions) context.Item = 

Also we noticed that each ForEach-objects gets called again when using looping like above. We solved it by ignoring already processed entries. And when reassigning the item to another user the old user is still the context.Item so we need to fetch it by index instead.

**Comment by ellinge** [comment]
If I check inside the Run-step. like here https://github.com/danielgerlag/workflow-core/issues/311. the context.Item is null as well so can't see how I can "extract it" again. Do I need a Approver-property on my workflowdata to store the current context.Item? I think I tried something like that but every step got the same ”user” then.

**Comment by danielgerlag** [comment]
I tested a simple example of this and could not reproduce the issue.

**Comment by danielgerlag** [comment]
https://github.com/danielgerlag/workflow-core/pull/683

**Comment by ellinge** [comment]
The issue seems to be resolved but I found another additional issue, probably related to the fix you applied for the context.Item, 
![image](https://user-images.githubusercontent.com/17863113/106966962-962f8380-6746-11eb-9bf6-0059183ed309.png)

We basically have multiple parallel tasks which we send an email to the each assigned user and they have a page which they can approve, reject or reassign the task. When reassigning the task we send an outcome of false. This does not seem to be picked up by or When(_ => false), or not sent to it. So workflow continues to the next batch of users (another role) instead of sending a new mail to the new user.

**Comment by ellinge** [comment]
Seems to be fixed now, but have another issue related to using When with outcomes and looping which I create a separate issue for. Thanks!
